### PR TITLE
[클린코드 7기 김진주] 로또 미션 STEP 2

### PR DESCRIPTION
**Pull request by devpearlkim**
안녕하세요, 태은님!
요구사항을 중심으로 구현했습니다
지난 스텝에서 피드백 해주신 부분들이 부분들이 큰 도움이 되었습니다. 
이번에도 핵심에서 벗어나거나 개선이 필요한 부분이 있다면 피드백 부탁드립니다.
감사합니다! 😊

### 구현 내용
당첨 통계 출력 후 재시작/종료 여부를 입력받도록 구현
잘못된 입력 시 throw를 통해 예외를 발생시키고, 에러 메시지 출력 후 재입력 처리

### 궁금한 점
다른 분의 코드를 보니 입력 단계에서 별도의 검증 없이 도메인에서 예외를 던지고, 재입력을 처리하는 방식으로 구현한 사례가 있었습니다.
반면, 저는 UI에서 validation 함수를 통해 사전에 검증하고 다시 입력받도록 구현했는데요.
이 두 접근법의 장단점에 대해 태은님은 어떻게 생각하시는지 의견이 궁금합니다!

Rebase 과정에서 이전 커밋이 함께 노출되었습니다.
Step2의 실제 구현 내용은 아래 커밋부터입니다. 번거로우시겠지만, 참고 부탁드립니다! 

👉 [feat: Add restart functionality using do-while loop in main](https://github.com/next-step/js-lotto/commit/62e37bf0a08f1cf9380cb514306a7834734a487a)

### CODE REVIEW

File: __tests__/validation.test.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  isValidPurchaseAmount,
+} from "../src/validation.js";
+
+describe("Validation Functions", () => {
+  const range = { min: 1, max: 45 };
+
+  describe("isValidNumberInRange", () => {
+    test("should return true for the minimum boundary value", () => {
+      expect(isValidNumberInRange(1, range)).toBe(true);
+    });
```
Comment (Tanney-102) [inline __tests__/validation.test.js:14]: range를 조금 더 명확한 맥락 안에서 선언하면 좋겠어요.

옵션1) test 콜백 안에 선언하여 어떤 환경에서 테스트가 이뤄지는지 충분히 예측 가능하도록 한다.
```ts
  test("should return true for the minimum boundary value", () => {
      const range = { min: 1, max: 45 }
      expect(isValidNumberInRange(1, range)).toBe(true);
    });
```

옵션2) describe를 통해 테스트 환경을 더 명확하게 명시 한다.
```ts
  describe("isValidNumberInRange: given range [1, 45]", () => {
    const range = { min: 1, max: 45 };

    test("should return true for the minimum boundary value", () => {
      expect(isValidNumberInRange(1, range)).toBe(true);
    });
  })
```

File: src/view/ui/common/input.js [hunk 2]
```diff
@@ -1,5 +1,24 @@
 import readline from "readline";
 
+export async function promptWithValidation({
+  query,
+  validationFn,
+  errorMessage,
+  transformFn = (input) => input,
```
Comment (Tanney-102) [inline src/view/ui/common/input.js:7]: 오 검증로직과 변환로직을 밖에서 주입하는 방식 좋네요 👍

**Review by Tanney-102 — APPROVED** [review]
안녕하세요 진주님!
빠르게 지난 리뷰 반영 및 restart 기능 구현까지 잘 해주셨네요! 
특별하게 문제될 만한 부분은 없었고 테스트 쪽에 하나 정도 코멘트 남겨두었으니 한 번 고민해보시면 좋겠습니다.

> 다른 분의 코드를 보니 입력 단계에서 별도의 검증 없이 도메인에서 예외를 던지고, 재입력을 처리하는 방식으로 구현한 사례가 있었습니다.
반면, 저는 UI에서 validation 함수를 통해 사전에 검증하고 다시 입력받도록 구현했는데요.
이 두 접근법의 장단점에 대해 태은님은 어떻게 생각하시는지 의견이 궁금합니다!

진주님도 동일하게 도메인 쪽에서도 검증을 하고 계시네요~!(물론 ui쪽에서 먼저 거르긴 하지만요)
어떤 값이 유효하지 유효하지 않는지도 결국 도메인 지식에 의존하는 경우가 많기 때문에 도메인에서 검증을 하는 건 당연해보이네요. 이 사실만 기본 전제로 깔고 간다면 나머지는 상황에 따라 판단 해야할 영역으로 보입니다. 몇가지 생각의 흐름만 예시로 남겨둘게요.

(생각1)
- 이미 도메인에서 검증을 하고 있는데 ui에서도 검증할 필요가 있을까? 
- 같은 검증을 양쪽에서 하니 그만큼 관리 비용도 늘고, UI에서 불필요하게 도메인을 많이 알게되는 것 같아서 좀 별론데
- UI는 in/out의 책임만 충실하게 수행하고 검증은 도메인에게 맡기자
- (or) UI가 일정부분 도메인을 의존하는 건 자연스러운 거 아님? 그냥 가자

(생각2)
- 검증에 있어서 UI와 도메인의 책임을 분리하는 것 꺄지는 좋음
- 근데 도메인과 별개로 UI에서 해야하는 검증은 없나?
- 빈값이 들어온다거나 type 자체가 다른 경우는 도메인까지 갈 필요도 없을 듯. 애초에 UI에서 막아버리자

(생각3)
- 콘솔 앱이 아니라 client(UI)와 server(도메인)가 있는 앱이라면?
- 도메인에서 검증을 하긴해야겠지만 UI에서 미리 검증을 해주면 불필요한 요청들을 막을 수 있지 않을까?
- 잘못된 입력으로 서버에 요청하는 빈도 + 이로인한 비용 vs client, server 양쪽에서 validation을 관리해야하는 비용을 비교해보자

**Comment by devpearlkim** [comment]
감사합니다, 태은님!
질문의 배경은 사실 제가 지금까지는 뷰와 도메인 로직을 분리해야 한다는 생각을 깊게 하지 못했기 때문에 어찌 보면 당연한 질문을 드렸던 것 같습니다.

"뷰에서는 도메인 로직을 감춰야 하는데? + 재입력은 필요하고? 그럼 뷰에서 검증해야 하나?"
이런 흐름으로 코드를 작성하면서도, 도메인에서만 검증하는 방법은 아예 떠올리지 못했던 것 같아요.

그런데 다른 분들의 코드를 보니 다양한 접근이 있어서, 혹시 일반적으로 선호되는 방식이나 분명한 장단점이 있을까 싶어 질문을 드렸던 거였는데, 질문이 조금 모호했던 것 같네요. 😅
상황에 따라 다르게 접근해야 하는 영역이라는 점과 각 관점에서의 생각을 자세히 설명해주셔서 큰 도움이 되었습니다!

다시 한 번 감사드립니다.😊
